metricflow/specs/merge_builder.py

```diff
@@ -0,0 +1,50 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Generic, Iterable, TypeVar
+
+SelfTypeT = TypeVar("SelfTypeT", bound="Mergeable")
+
+
+class Mergeable(ABC):
+    """Collections that can be merged together."""
+
+    @abstractmethod
+    def merge(self: SelfTypeT, other: SelfTypeT) -> SelfTypeT:
+        """Return a new object that is the result of merging self with other."""
+        pass
+
+
+MergeableT = TypeVar("MergeableT", bound="Mergeable")
+
+
+class MergeBuilder(Generic[MergeableT]):
+    """Implementation of the builder pattern for merging many mergeable objects."""
+
+    def __init__(self, initial_item: MergeableT) -> None:  # noqa: D
+        self._accumulator = initial_item
+
+    def add(self, other: MergeableT) -> None:  # noqa: D
+        if self._accumulator is None:
```

Comment:
Do we need this? Can the generic actually be `Optional[MergeableThing]`? I don't think it can, with a bound of "Mergeable" here.